mail_allow_portal_internal_note/models/mail_message.py

```diff
@@ -0,0 +1,17 @@
+# Copyright 2021 Open Source Integrators
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import models
+
+
+class Message(models.Model):
+    _inherit = "mail.message"
+
+    def _get_search_domain_share(self):
+        if (
+            self.user_has_groups("base.group_portal")
+            and self.env.user.special_portal_users
+        ):
+            return []
+        else:
+            return super(Message, self)._get_search_domain_share()
```

Comment:
Make this based on having a Group, instead of a checkbox.
Also, it is best to always call super(), to avoid breaking the inheritance chain.
```suggestion
        domain = super(Message, self)._get_search_domain_share()
        if self.user_has_groups("mail_allow_portal_internal_note.see_internal_notes"):
            domain = []
        return domain
```